module-system/module-implementations/sov-nft-module/src/offchain.rs

```diff
@@ -0,0 +1,168 @@
+                            &metadata_url,
+                            &(supply as i64),
+                        ],
+                    );
+                    if let Err(e) = result {
+                        println!("Failed to execute query: {}", e);
+                    }
+                }
+                Err(e) => {
+                    println!("Failed to connect to the database: {}", e);
+                }
+            }
+        } else {
+            println!("Environment variable POSTGRES_CONNECTION_STRING is not set");
+        }
+    })
+}
+
+// CREATE TABLE nft (
+//     collection_address TEXT NOT NULL,
+//     nft_id BIGINT NOT NULL,
+//     metadata_url TEXT,
+//     owner TEXT NOT NULL,
+//     frozen BOOLEAN NOT NULL,
+//     PRIMARY KEY (collection_address, nft_id)
+// );
+
+#[offchain]
+pub fn track_nft<C: sov_modules_api::Context>(nft: &Nft<C>) {
+    let collection_address = nft.get_collection_address().to_string();
+    let nft_id = nft.get_token_id();
+    let owner = nft.get_owner().to_string();
+    let frozen = nft.is_frozen();
+    let metadata_url = nft.get_token_uri();
+    tokio::task::block_in_place(|| {
+        if let Ok(conn_string) = std::env::var("POSTGRES_CONNECTION_STRING") {
+            match postgres::Client::connect(&conn_string, NoTls) {
+                Ok(mut client) => {
+                    let result = client.execute(
+                        "INSERT INTO nfts (\
+                        collection_address, nft_id, metadata_url,\
+                        owner, frozen)\
+                        VALUES ($1, $2, $3, $4, $5)\
+                        ON CONFLICT (collection_address, nft_id)\
+                        DO UPDATE SET metadata_url = EXCLUDED.metadata_url,\
+                                      owner = EXCLUDED.owner,\
+                                      frozen = EXCLUDED.frozen",
+                        &[
+                            &collection_address,
+                            &(nft_id as i64),
+                            &metadata_url,
+                            &owner,
+                            &frozen,
+                        ],
+                    );
+                    if let Err(e) = result {
+                        println!("Failed to execute query: {}", e);
+                    }
+                }
+                Err(e) => {
+                    println!("Failed to connect to the database: {}", e);
+                }
+            }
+        } else {
+            println!("Environment variable POSTGRES_CONNECTION_STRING is not set");
+        }
+    })
+}
+
+// CREATE TABLE top_owners (
+//     owner TEXT NOT NULL,
+//     collection_address TEXT NOT NULL,
+//     count BIGINT NOT NULL,
+//     PRIMARY KEY (owner, collection_address)
+// );
+
+#[offchain]
+pub fn update_top_owners<C: sov_modules_api::Context>(
+    collection_address: &CollectionAddress<C>,
+    owner_count_incr: Option<&[(String, u64)]>,
+    owner_count_decr: Option<&[(String, u64)]>,
```

Comment:
Question: why do we use two vecs instead of using a signed integer?

Reply:
ah yeah, this logic is changed now with merging update_nft and update_top_owners (original reason was to decerement counts of owner1 and increment counts of owner2 for a transfer, but that logic doesnt matter right now)